t/startfile.lua

```diff
@@ -0,0 +1,280 @@
+    -- does the hash selector hold a reference to the pool (but only available in main config?)
+
+    -- uncomment to use the ketama loadable module.
+    -- FIXME: passing an argument to the ketama module doesn't work yet.
+    -- local ketama = require("ketama")
+
+    -- convert the pools into hash selectors.
+    -- TODO: is this a good place to add prefixing/hash editing?
+    for _, subs in pairs(main_zones) do
+        for k, v in pairs(subs) do
+            -- use next line instead for a third party ketama hash
+            -- subs[k] = mcp.pool(v, { dist = ketama })
+            -- this line overrides the default bucket size for ketama
+            -- subs[k] = mcp.pool(v, { dist = ketama, obucket = 80 })
+            -- this line uses the default murmur3 straight hash.
+            subs[k] = mcp.pool(v)
+
+            -- use this next line instead for jump hash.
+            -- the order of servers in the pool argument _must_ not change!
+            -- adding the seed string will give a different key distribution
+            -- for each zone.
+            -- NOTE: 'k' may not be the right seed here:
+            -- instead stitch main_zone's key + the sub key?
+            -- subs[k] = mcp.pool(v, { dist = mcp.hash_jump, seed = k })
+        end
+    end
+
+    return main_zones
+end
+
+-- WORKER CODE:
+
+-- need to redefine main_zones using fetched selectors?
+
+-- TODO: Fallback zone here?
+function failover_factory(zones, local_zone)
+    local near_zone = zones[local_zone]
+    local far_zones = {}
+    -- NOTE: could shuffle/sort to re-order zone retry order
+    -- or use 'next(far_zones, idx)' via a stored upvalue here
+    for k, v in pairs(zones) do
+        if k ~= local_zone then
+            far_zones[k] = v
+        end
+    end
+    return function(r)
+        local res = near_zone(r)
+        if res:hit() == false then
+            for _, zone in pairs(far_zones) do
+                res = zone(r)
+                if res:hit() then
+                    break
+                end
+            end
+        end
+        return res -- send result back to client
+    end
+end
+
+-- SET's to main zone, issues deletes to far zones.
+function setinvalidate_factory(zones, local_zone)
+    local near_zone = zones[local_zone]
+    local far_zones = {}
+    -- NOTE: could shuffle/sort to re-order zone retry order
+    -- or use 'next(far_zones, idx)' via a stored upvalue here
+    for k, v in pairs(zones) do
+        if k ~= local_zone then
+            far_zones[k] = v
+        end
+    end
+    local new_req = mcp.request
+    return function(r)
+        local res = near_zone(r)
+        if res:ok() == true then
+            -- create a new delete request
+            local dr = new_req("delete /testing/" .. r:key() .. "\r\n")
+            for _, zone in pairs(far_zones) do
+                -- NOTE: can check/do things on the specific response here.
+                zone(dr)
```

Comment:
Is there a way to do the deletes asynchronously, e.g. send the SET response to the client before sending a delete.

If the duration of the far zones is high (e.g. 10ms vs 0.1ms near) this might impact the performance of an application using the library

(e.g. something analogous to process.nextTick/async  in JavaScript, e.g. provided through the C library)

Also, it's potentially confusing to readers to use `zone` as a local variable if it's also a global function

Reply:
Oh, right, there's the `noreply` option on all of the meta-text commands for common use cases or approaches. There's still questions such as whether the lua library should inject some no-ops mixed with noreply to make that easier, etc.

The ability to do async work after sending a response might still be useful for some replication use cases, e.g. to only delete the key from the remote datacenter after `GET`ting the value and seeing that it exists but is not the same (or has different flags)

- mcrouter does that in some of its configurations for pools, e.g. there's an option to send the first successful response
- those approaches might run into issues with memory if the far zone is unresponsive, having networking issues, is not acknowledging or rejecting packets, is hitting a networking bottleneck, or is just far away, but that's something I assume should be focused on much later rather than right now

Reply:
A simpler approach to `async` may be an API such as `r.send_early_response(res)`, and subsequent calls (or the return value) would get ignored (and freed if needed)

Reply:
I don't think `zone` is a global variable... I'll check though. Keep in mind at least `t/startfile.lua` is mostly there for the test suite and I'm just making things up :) Over the last few weeks I've been simplifying the terminology (ie; "hash selectors" are now just... pools, duh.). Appreciate pointers here.

RE: async stuff, check out the `mcp.await` interface... which I didn't apparently write many examples for. startfile.lua has walkallfactory using it.

The interface is: `mcp.await(request, table_of_pools, wait_res_count)` - wait_res_count is optional, defaults to length of the table. In its first version you can take a single request and issue parallel backend requests to N pools, waiting for the 1-n "good responses" before resuming the function. Main limitation to start is that you can't call functions, just pools directly. You also can't control the status result it's looking for (which is easier to fix).

In your example with SET's and delete's, you can do something like:
- execute the set
- once the set response is received, `mcp.await(new_delete_request, other_zone_pools, 0)`, which should immediately resume the corutine and asynchronously execute deletes.
- return result object from the original set

In the future it should be expandable to build an await context with multiple requests so you can issue that all at once, but for now this does actually work.

Reply:
Note await can be helpful for get failover situations as well: if the initial local zone get fails, you can issue parallel requests and take the winner, using a bit more backend resources to avoid making the client wait a potentially long time.

Reply:
> Ah I see you did want to get the result of a `get` or `set` first. In that case this does what you want. If you want to do further processing after early-returning a result, that's not possible yet. It might have to get implemented as returning a `res, function` instead of returning `res`, then function would get called as a continuation. Will see if I get a better idea later.

I had missed the part about `await` existing, but still expect some use cases such as the one I mentioned would benefit from async.

The `res, function` seems appropriate with lua's lexical scope of closures/functions.

Reply:
I think async/await are the same thing conceptually. It's in a minimal form now but internally it's creating an await object, stacking some stuff on it, then suspending the coroutine. Once I get a better sense for the API, those steps can be split out; create the await object in lua, stack things on it in a loop, then execute it later (or cancel) if you want. I had it set up like this originally but am trying to minimize API churn and kept it to the simplest interface to start.

`res, function` is more of a side effect of how I wrote things... Thankfully lua gives us a lot of options here.